fix torch.cat bug with boxed CPUFallback

bdhirsh opened the pull request:
Fixes https://github.com/pytorch/pytorch/issues/60902

The boxed fallback was written to assume that there was at least one tensor argument, which it used to figure out what device to move the cpu tensors to. That fails with an op like `torch.cat()`, which doesn't have any tensor arguments, but instead has a single `TensorList` argument. It looks like that wasn't caught originally because there weren't any ops that xla tests CPU fallbacks for that have only TensorList args.

I also added handling to gracefully deal with the case where you have an empty list of tensors - in that case we don't know what device to move everything to, but that doesn't matter because an empty list of tensors implies that we have no tensors to move anyway.

I tested it out though and noticed that `torch.cat(())` doesn't handle empty lists well anyway (erroring out in the dispatcher). I'm not sure that it's a huge issue, and not even sure that we want to fix it (default to CPU? add an extra codegen'd check into every op that only takes TensorList args?) but I'll file a separate bug for that: https://github.com/pytorch/pytorch/issues/60997

I tested it by running the pytorch/xla suite after removing `cat` from `xla_native_functions.yaml`, and confirming that we don't segfault anymore.

Stack from [ghstack](https://github.com/ezyang/ghstack):
* **#60993 fix torch.cat bug with boxed CPUFallback**

Differential Revision: [D29471577](https://our.internmc.facebook.com/intern/diff/D29471577)

bdhirsh committed:
fix torch.cat bug with boxed CPUFallback

[ghstack-poisoned]

bdhirsh commented:
@bdhirsh has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D29471577).

asuhan commented on aten/src/ATen/native/CPUFallback.cpp:
```diff
@@ -135,18 +141,46 @@ void cpu_fallback(const c10::OperatorHandle& op, torch::jit::Stack* stack) {
+          } else {
+            // We need to loop through all of the (potentially multiple) TensorList arguments
+            // In case, e.g. the first one is empty but the second is not.
+            for (auto& tens_list : tensorlist_args) {
```
This should go in a helper, it's too hard to tell right away what these nested loops are doing.

bdhirsh committed:
Update on "fix torch.cat bug with boxed CPUFallback"

Fixes https://github.com/pytorch/pytorch/issues/60902

The boxed fallback was written to assume that there was at least one tensor argument, which it used to figure out what device to move the cpu tensors to. That fails with an op like `torch.cat()`, which doesn't have any tensor arguments, but instead has a single `TensorList` argument. It looks like that wasn't caught originally because there weren't any ops that xla tests CPU fallbacks for that have only TensorList args.

I also added handling to gracefully deal with the case where you have an empty list of tensors - in that case we don't know what device to move everything to, but that doesn't matter because an empty list of tensors implies that we have no tensors to move anyway.

I tested it out though and noticed that `torch.cat(())` doesn't handle empty lists well anyway (erroring out in the dispatcher). I'm not sure that it's a huge issue, and not even sure that we want to fix it (default to CPU? add an extra codegen'd check into every op that only takes TensorList args?) but I'll file a separate bug for that: https://github.com/pytorch/pytorch/issues/60997

I tested it by running the pytorch/xla suite after removing `cat` from `xla_native_functions.yaml`, and confirming that we don't segfault anymore.


Differential Revision: [D29471577](https://our.internmc.facebook.com/intern/diff/D29471577)

[ghstack-poisoned]

bdhirsh commented:
@bdhirsh has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D29471577).

asuhan commented on aten/src/ATen/native/CPUFallback.cpp:
```diff
@@ -39,6 +40,25 @@ std::vector<at::Tensor> to_cpu(const at::TensorList& tensors) {
+  // Decide what device to move the output tensor(s) to.
+  // The current convention is that we use the first tensor arg to pick the device
+  // Barring that, we take the first tensor from a TensorList arg.
+  c10::optional<c10::Device> tgt_device = c10::nullopt;
```
This is unused.

bdhirsh committed:
Update on "fix torch.cat bug with boxed CPUFallback"

Fixes https://github.com/pytorch/pytorch/issues/60902

The boxed fallback was written to assume that there was at least one tensor argument, which it used to figure out what device to move the cpu tensors to. That fails with an op like `torch.cat()`, which doesn't have any tensor arguments, but instead has a single `TensorList` argument. It looks like that wasn't caught originally because there weren't any ops that xla tests CPU fallbacks for that have only TensorList args.

I also added handling to gracefully deal with the case where you have an empty list of tensors - in that case we don't know what device to move everything to, but that doesn't matter because an empty list of tensors implies that we have no tensors to move anyway.

I tested it out though and noticed that `torch.cat(())` doesn't handle empty lists well anyway (erroring out in the dispatcher). I'm not sure that it's a huge issue, and not even sure that we want to fix it (default to CPU? add an extra codegen'd check into every op that only takes TensorList args?) but I'll file a separate bug for that: https://github.com/pytorch/pytorch/issues/60997

I tested it by running the pytorch/xla suite after removing `cat` from `xla_native_functions.yaml`, and confirming that we don't segfault anymore.


Differential Revision: [D29471577](https://our.internmc.facebook.com/intern/diff/D29471577)

[ghstack-poisoned]

bdhirsh commented:
@bdhirsh has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D29471577).

ezyang reviewed (approved):
okey dokey

bdhirsh committed:
Update on "fix torch.cat bug with boxed CPUFallback"

Fixes https://github.com/pytorch/pytorch/issues/60902

The boxed fallback was written to assume that there was at least one tensor argument, which it used to figure out what device to move the cpu tensors to. That fails with an op like `torch.cat()`, which doesn't have any tensor arguments, but instead has a single `TensorList` argument. It looks like that wasn't caught originally because there weren't any ops that xla tests CPU fallbacks for that have only TensorList args.

I also added handling to gracefully deal with the case where you have an empty list of tensors - in that case we don't know what device to move everything to, but that doesn't matter because an empty list of tensors implies that we have no tensors to move anyway.

I tested it out though and noticed that `torch.cat(())` doesn't handle empty lists well anyway (erroring out in the dispatcher). I'm not sure that it's a huge issue, and not even sure that we want to fix it (default to CPU? add an extra codegen'd check into every op that only takes TensorList args?) but I'll file a separate bug for that: https://github.com/pytorch/pytorch/issues/60997

I tested it by running the pytorch/xla suite after removing `cat` from `xla_native_functions.yaml`, and confirming that we don't segfault anymore.


Differential Revision: [D29471577](https://our.internmc.facebook.com/intern/diff/D29471577)

[ghstack-poisoned]
